### PR TITLE
[TACHYON-644] Create UserCleanup thread to remove zombie users

### PR DESCRIPTION
**Pull request by miasantreble**
https://tachyon.atlassian.net/browse/TACHYON-644

### CODE REVIEW

**Comment by apc999** [comment]
@calvinjia please review this PR

File: servers/src/main/java/tachyon/worker/block/BlockWorker.java [hunk 1]
```diff
@@ -129,15 +131,18 @@ public BlockWorker(TachyonConf tachyonConf) throws IOException {
             mStartTimeMs, mTachyonConf);
 
     // Setup Worker to Master Syncer
-    // We create two threads for two syncers: mBlockMasterSync and mPinListSync
+    // We create three threads for three syncers: mBlockMasterSync, mUserCleanup and mPinListSync
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockWorker.java]: "three syncers" are not accurate. It should be two syncers and one cleaner(??)

File: servers/src/main/java/tachyon/worker/block/UserCleanup.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import tachyon.conf.TachyonConf;
+import tachyon.util.CommonUtils;
+
+/**
+ * UserCleanup periodically checks if any user have become zombies,
+ * removes the zombie user and associated data when necessary
+ * The syncing parameters (intervals) adopt directly from worker-to-master heartbeat
+ * configurations.
+ */
+public class UserCleanup implements Runnable {
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/UserCleanup.java]: NIT: How about naming this class to something like `zombieUserCleaner` or maybe just `userCleaner`, to be more consistent with current naming convention?

File: servers/src/main/java/tachyon/worker/block/UserCleanup.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import tachyon.Constants;
+import tachyon.conf.TachyonConf;
+import tachyon.util.CommonUtils;
+
+/**
+ * UserCleanup periodically checks if any user have become zombies,
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/UserCleanup.java]: Please reformat this javadoc.

File: servers/src/main/java/tachyon/worker/block/BlockWorker.java [hunk 4]
```diff
@@ -129,15 +131,18 @@ public BlockWorker(TachyonConf tachyonConf) throws IOException {
...
-        Executors.newFixedThreadPool(2, ThreadFactoryUtils.build("worker-heartbeat-%d", true));
+        Executors.newFixedThreadPool(3, ThreadFactoryUtils.build("worker-heartbeat-%d", true));
     mBlockMasterSync = new BlockMasterSync(mBlockDataManager, mTachyonConf, mWorkerNetAddress);
     mBlockMasterSync.registerWithMaster();
 
     // Setup PinListSyncer
     mPinListSync = new PinListSync(mBlockDataManager, mTachyonConf);
 
+    // Setup UserCleanup
+    mUserCleanup = new UserCleanup(mBlockDataManager, mTachyonConf);
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/BlockWorker.java]: Maybe suffix this with `Thread` to be more descriptive.

**Comment by calvinjia** [comment]
@miasantreble thanks for the patch, looks good overall just some comments about the naming.

**Comment by miasantreble** [comment]
@calvinjia the PR didn't cause any regression so I'm not sure if we have test coverage for automatic zombie user cleanup. Are you talking about this test? https://github.com/amplab/tachyon/blob/f37654468d8693f491c27c069bc64da5c91b0a79/servers/src/test/java/tachyon/worker/block/BlockMetadataManagerTest.java#L237 but don't think this test go through the same code path. Please let me know if there are other test cases that need updating, thanks

**Comment by calvinjia** [comment]
I was thinking since this has now become an isolated component, does it make sense to unit test the user cleaner?

File: servers/src/main/java/tachyon/worker/block/BlockWorker.java (hunk 1 above)
Comment (miasantreble) [inline servers/src/main/java/tachyon/worker/block/BlockWorker.java]: done

File: servers/src/main/java/tachyon/worker/block/UserCleanup.java (hunk 2 above)
Comment (miasantreble) [inline servers/src/main/java/tachyon/worker/block/UserCleanup.java]: changed to UserCleaner, done

File: servers/src/main/java/tachyon/worker/block/UserCleanup.java (hunk 3 above)
Comment (miasantreble) [inline servers/src/main/java/tachyon/worker/block/UserCleanup.java]: Done

File: servers/src/main/java/tachyon/worker/block/BlockWorker.java (hunk 4 above)
Comment (miasantreble) [inline servers/src/main/java/tachyon/worker/block/BlockWorker.java]: Done

**Comment by miasantreble** [comment]
@calvinjia do you think the test for UserCleaner can be added later in a separate JIRA? I checked other threads, looks like we don't have lots of test coverage for threads overall(PinListSync, BlockMasterSync, NIODataServer, etc), maybe we can create a new JIRA to add the tests for them altogether?

**Comment by calvinjia** [comment]
+1 LGTM

**Comment by apc999** [comment]
+1 LGTM

**Pull request merged**